File: 3.x.x/crates-utopia/query/src/builder/parsed_query.rs
//! PHP `Utopia\Query\Builder\ParsedQuery`.

use crate::enums::CursorDirection;
use crate::query::Query;
use crate::value::QueryValue;

#[derive(Debug, Clone, PartialEq, Default)]
pub struct TimeBucket {
    pub attribute: String,
    pub interval: String,
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct ParsedQuery {
    pub filters: Vec<Query>,
    pub selections: Vec<Query>,
    pub aggregations: Vec<Query>,
    pub group_by: Vec<String>,
    pub having: Vec<Query>,
    pub distinct: bool,
    pub joins: Vec<Query>,
    pub unions: Vec<Query>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub cursor: Option<QueryValue>,
    pub cursor_direction: Option<CursorDirection>,
    pub time_buckets: Vec<TimeBucket>,
}

File: packages/query/src/Builder/ParsedQuery.php
<?php

namespace Utopia\Query\Builder;

use Utopia\Query\CursorDirection;
use Utopia\Query\Query;

readonly class ParsedQuery
{
    /**
     * @param  list<Query>  $filters
     * @param  list<Query>  $selections
     * @param  list<Query>  $aggregations
     * @param  list<string>  $groupBy
     * @param  list<Query>  $having
     * @param  list<Query>  $joins
     * @param  list<Query>  $unions
     * @param  list<array{attribute: string, interval: string}>  $timeBuckets
     */
    public function __construct(
        public array $filters = [],
        public array $selections = [],
        public array $aggregations = [],
        public array $groupBy = [],
        public array $having = [],
        public bool $distinct = false,
        public array $joins = [],
        public array $unions = [],
        public ?int $limit = null,
        public ?int $offset = null,
        public mixed $cursor = null,
        public ?CursorDirection $cursorDirection = null,
        public array $timeBuckets = [],
    ) {
    }
}
